### PR TITLE
CETS support for S2S

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR addresses MIM-1931

Proposed changes include:
* Add backend module
* Both mnesia and cets are supported
* Added `s2s/` directory, because there are a lot of s2s files now (total 6). We could also move ejabberd_s2s, ejabberd_s2s_in, ejabberd_s2s_out into this directory later, so all s2s files are located together (not in this PR, because it would make git merging of the feature branch harder). 

Changes after the first review:
* Use `ejabberd_s2s:fromto()` type in a lot of places where possible. It simplifies logic a bit (one less argument to pass too).
* Move logic functions into `mongoose_s2s_lib`. `ejabberd_s2s` module is pretty big already, so separating logic functions and functions that make side effects (i.e. routing, opening connections and writing data across the cluster) helps to understand code faster.
* There are limited number of exported functions from `mongoose_s2s_lib`, so the reader don't need to worry about all extra functions when trying to understand the logic for opening s2s connection.
* Handle joining cluster with conflicting secrets working with CETS. There is a new option `handle_conflict` which helps to choose which record to keep, when record is on both nodes that are trying to join together https://github.com/esl/cets/pull/4
* Added comments to dialback functions, use exml in some places instead of deprecated xml functions.
* Removed `ejabberd_s2s_out:stop_connection` call `from ejabberd_s2s_in`. There is a comment about it. If we want to stop a connection, it should not be just one connection. But probably we should not stop connections from `ejabberd_s2s_in`
* Made a module with dialback steps mongoose_s2s_dialback - because it is extremely hard to follow dialback logic with the format for packets defined in https://xmpp.org/extensions/xep-0220.html (same tag is reused for key sending and dialback result, i.e. step1 and step4). While db:result for step1 makes no sense at all... So, at least step numbers kinda make sense if you also keep the diagram open.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/179125) / small_tests / 1bf1dcc0003ecaaeecee30c326c1729ccc54a549
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/179125/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/179125/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-29_14.14.22/lib.mongooseim.logs/run.2023-06-29_14.14.23/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/179155) / small_tests / 0f3819bd97cbb3f1168763acc5d59bcd0c0804b9
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/179155/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/179155/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-06-29_16.35.26/lib.mongooseim.logs/run.2023-06-29_16.35.27/suite.log.html)

File: src/ejabberd_s2s.erl [hunk 1]
```diff
@@ -614,3 +528,41 @@ lookup_certfile(HostType) ->
...
+                     stacktrace => Stacktrace}),
+         {error, Reason}
+    end.
+
+%% Returns true if the connection is registered
+-spec call_try_register(Pid :: pid(), ShouldWriteF :: fun(), FromTo :: fromto()) -> boolean().
+call_try_register(Pid, ShouldWriteF, FromTo) ->
+    mongoose_s2s_backend:try_register(Pid, ShouldWriteF, FromTo).
+
+call_node_cleanup(Node) ->
```
Comment (chrzaszcz) [inline src/ejabberd_s2s.erl:556]: A bit inconsistent - some functions have specs, some don't.

**Review by chrzaszcz — COMMENTED** [review]
It looks good in general. I added some comments.

File: src/config/mongoose_config_spec.erl [hunk 2]
```diff
@@ -210,6 +213,7 @@ general_defaults() ->
       <<"language">> => <<"en">>,
       <<"all_metrics_are_global">> => false,
       <<"sm_backend">> => mnesia,
+      <<"s2s_backend">> => mnesia,
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/config/mongoose_config_spec.erl:216]: there were some issues that the option is seen as per host_type.
I.e. to access it I need to use:
```
mongoose_config:get_opt([{s2s, HostType}, max_retry_delay])
```

Probably should've invistigate how `wrap = global_config` works more, but it was failing, so I've just made s2s_backend option. There would be also component_backend option anyway. And to configure them you still need to edit them in one place anyway. Still, I could rework to s2s.backend once I see the complete picture with all the modules converted :)

File: src/ejabberd_s2s.erl [hunk 3]
```diff
@@ -614,3 +528,41 @@ lookup_certfile(HostType) ->
...
+
+db_init() ->
+    Backend = mongoose_config:get_opt(s2s_backend),
+    mongoose_s2s_backend:init(#{backend => Backend}).
+
+-spec dirty_read_s2s_list_pids(FromTo :: fromto()) -> {ok, [pid()]} | {error, Reason :: term()}.
+dirty_read_s2s_list_pids(FromTo) ->
+    try
+        mongoose_s2s_backend:dirty_read_s2s_list_pids(FromTo)
+    catch Class:Reason:Stacktrace ->
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/ejabberd_s2s.erl:543]: get_connections_pids uses an empty list, if the code crashes.
find_connection returns an error as a tuple (but do not crashes).

So, try catch here so the caller could handle both exceptions and `{error,_}` tuple with the same code.

File: src/config/mongoose_config_spec.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/config/mongoose_config_spec.erl:216]: S2S section is indeed wrapped in `host_config`. In this case let's keep `s2s_backend`.

File: src/ejabberd_s2s.erl (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/ejabberd_s2s.erl:543]: I understand the purpose of wrapping the error, but I think we should only catch expected errors. So waht kind of error do we expect from a dirty read?

File: src/s2s/mongoose_s2s_mnesia.erl [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+-record(s2s, {
+          fromto :: ejabberd_s2s:fromto() | '_',
+          pid :: pid() | '$1'
+         }).
+
+-record(s2s_secret, {host_type, source, secret}).
+
+-include("mongoose_logger.hrl").
+
+init(_) ->
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/s2s/mongoose_s2s_mnesia.erl:22]: Yes, only for the reader, but they would help a lot.

File: src/mongoose_listener_sup.erl [hunk 5]
```diff
@@ -16,7 +16,8 @@ start_link() ->
 
 -spec start_child(supervisor:child_spec()) -> ok.
 start_child(ChildSpec) ->
-    {ok, _Pid} = supervisor:start_child(?MODULE, ChildSpec),
+    Res = supervisor:start_child(?MODULE, ChildSpec),
+    check_start_child_result(Res, ChildSpec),
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/mongoose_listener_sup.erl:20]: It would add ChildSpec into the error, not just a badmatch.
So, we would know which listener is crashing.

File: src/ejabberd_s2s.erl [hunk 6]
```diff
@@ -99,89 +91,58 @@ filter(From, To, Acc, Packet) ->
...
-        end,
-    Res = mnesia:async_dirty(F),
+    Res = call_node_cleanup(Node),
     {ok, maps:put(?MODULE, Res, Acc)}.
 
 -spec key(mongooseim:host_type(), {jid:lserver(), jid:lserver()}, binary()) ->
     binary().
 key(HostType, {From, To}, StreamID) ->
-    Secret = get_shared_secret(HostType),
+    {ok, {_, Secret}} = get_shared_secret(HostType),
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/ejabberd_s2s.erl:145]: It is used in `set_shared_secret`.
There is some logic that decides when to generate a random secret, if not specified in the config (and when to overwrite the secret, if specified in the config). Kinda complicated to begin with, just adding Source makes code a bit more understandable in that function.

File: src/ejabberd_s2s.erl (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/ejabberd_s2s.erl:543]: Ok let's keep it

File: src/ejabberd_s2s.erl [hunk 7]
```diff
@@ -385,57 +306,33 @@ choose_pid(From, Pids) ->
...
-            hd(ConnectionsResult);
-        PIDs ->
-            {atomic, choose_pid(From, PIDs)}
-    end.
+    Server :: jid:server(), FromTo :: fromto()) -> ok.
+open_several_connections(N, MyServer, Server, FromTo) ->
+    ShouldWriteF = should_write_f(FromTo),
+    [new_connection(MyServer, Server, FromTo, ShouldWriteF)
+     || _N <- lists:seq(1, N)],
+    ok.
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/ejabberd_s2s.erl:314]: Do you mean we have `ShouldWriteF` now here?
There was still `mnesia:read({s2s, FromTo}),` and `needed_connections_number` call in the `new_connection` old implementation (i.e. a table lookup).

This `ShouldWriteF` helps to abstract `needed_connections_number` logic away from the backend module.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/179897) / small_tests / e7bbf38d736d74d588f80ab1325d5853d29f70bd
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/179897/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/179897/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-07-07_07.41.15/lib.mongooseim.logs/run.2023-07-07_07.41.15/suite.log.html)

File: src/mongoose_listener_sup.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/mongoose_listener_sup.erl:20]: Technically we would get some info from the SASL about the the child failing to start.
But it is still useful to see ChildSpec in `{ok, Pid}` badmatch.

File: src/ejabberd_s2s.erl (hunk 6 above)
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/ejabberd_s2s.erl]: It is removed now.

File: src/ejabberd_s2s.erl (hunk 7 above)
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/ejabberd_s2s.erl:224]: Rewrote the logic, no retries here now (and no ShouldWriteF, but we have now a call to a library function from the backend modules:

```erlang
try_register(Pid, FromTo) ->
    Pids = get_s2s_out_pids(FromTo),
    case mongoose_s2s_lib:need_more_connections(FromTo, Pids) of
        true ->
            cets:insert(?TABLE, {{FromTo, Pid}}),
            true;
        false ->
            false
    end.
  ```
  
  `mongoose_s2s_lib:need_more_connections` is the call.

File: src/ejabberd_s2s.erl [hunk 8]
```diff
@@ -360,18 +268,31 @@ maybe_open_several_connections(From, To, MyServer, Server, FromTo,
...
+    NeededConnections = needed_connections_number(
+                          L, MaxS2SConnectionsNumber,
+                          MaxS2SConnectionsNumberPerNode),
+    case NeededConnections > 0 of
+        true ->
+            %% We establish the missing connections for this pair.
+            open_several_connections(
+              NeededConnections, MyServer,
+              Server, FromTo),
+            find_connection(From, To, Retries - 1);
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/ejabberd_s2s.erl]: Removed retry mechanism, it was more or less code reuse.
Now rewrote to have two separate steps: 
- one ensures that there are registered connections.
- and the second one chooses pid from the list of the registered connections.
It reduces nesting a bit and we don't need two functions (one to open brand new connections from the first time and the second is to open missing connections - now it is just "open number of connections").

**Comment by mongoose-im** [comment]
[small_tests_25_arm64](https://circleci.com/gh/esl/MongooseIM/180013) / small_tests / 27d2c6130bc14b09f094a2b5b524e52ea0c42a35
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/180013/small_tests.25.3-arm64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/180013/small_tests.25.3-arm64/small/ct_run.mongooseim@localhost.2023-07-07_15.19.33/lib.mongooseim.logs/run.2023-07-07_15.19.34/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/180043) / small_tests / 8d4329be38070bb87523b66096df3944376191eb
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/180043/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/180043/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-07-07_18.52.02/lib.mongooseim.logs/run.2023-07-07_18.52.02/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/180069) / small_tests / b479dbe028262e6dd22d647510ed09855d214bb6
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/180069/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/180069/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-07-07_19.42.06/lib.mongooseim.logs/run.2023-07-07_19.42.06/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/180100) / small_tests / e567c7341963e39885f31453245516a31f2fe002
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/180100/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/180100/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-07-07_22.29.55/lib.mongooseim.logs/run.2023-07-07_22.29.56/suite.log.html)

File: src/mongoose_listener_sup.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/mongoose_listener_sup.erl:20]: I could use export function from ejabberd_sup:start_child and use it for logging.
I still like some reporting from start_child.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/180313) / small_tests / 26a517b38b9b21d2ab4a65582c03c806c6b9a654
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/180313/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/180313/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-07-10_23.05.31/lib.mongooseim.logs/run.2023-07-10_23.05.31/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/180342) / small_tests / 18ef1d7c431e8f5cb26f908dcda973e3716af68b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/180342/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/180342/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-07-10_23.34.46/lib.mongooseim.logs/run.2023-07-10_23.34.47/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/180401) / small_tests / 3ab8d27bd7f61e98f1d633d1ab2dde94b0cfddc3
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/180401/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/180401/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-07-11_07.52.18/lib.mongooseim.logs/run.2023-07-11_07.52.18/suite.log.html)

File: src/s2s/mongoose_s2s_cets.erl [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    %% Non-random, non-node-specific keys
+    %% This means that default merging would not work
+    cets:start(?SECRET_TABLE, #{handle_conflict => fun ?MODULE:handle_secret_conflict/2}),
+    cets_discovery:add_table(mongoose_cets_discovery, ?TABLE),
+    cets_discovery:add_table(mongoose_cets_discovery, ?SECRET_TABLE),
+    ok.
+
+%% Store the most recent value:
+%% - first element of the tuple is the same and it is the key.
+%% - second element is a timestamp, so comparing tuples works.
```
Comment (pawlooss1) [inline src/s2s/mongoose_s2s_cets.erl:33]: Maybe instead of this two lines of comment a type spec would be sufficient?

File: src/s2s/mongoose_s2s_dialback.erl
Comment (pawlooss1) [inline src/s2s/mongoose_s2s_dialback.erl:1]: We should add `-xep([{xep, 220}, {version, "1.1.1"}]).` attribute to this module since it implements this XEP.

File: src/s2s/mongoose_s2s_info.erl [hunk 10]
```diff
@@ -0,0 +1,46 @@
+%% Some ugly code only used in tests.
```
Comment (chrzaszcz) [inline src/s2s/mongoose_s2s_info.erl:1]: I think that we should either leave it in `ejabberd_s2s` or rewrite it as less ugly code, and then put it in `mongoose_s2s_info`. We don't just move code and rename to `mongooseim`.

File: src/s2s/mongoose_s2s_lib.erl [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+%% Checks:
+%% - if the host is not a service
+%% - and if the s2s host is not blacklisted or is in whitelist
+-spec is_s2s_allowed_for_host(fromto(), _OldConnections :: s2s_pids()) -> boolean().
+is_s2s_allowed_for_host(_FromTo, [_|_]) ->
+    true; %% Has outgoing connections established, skip the check
+is_s2s_allowed_for_host(FromTo, []) ->
+    not is_service(FromTo) andalso allow_host(FromTo).
+
+%% Check if host is in blacklist or white list
```
Comment (chrzaszcz) [inline src/s2s/mongoose_s2s_lib.erl:148]:
```suggestion
%% Check if host is in blacklist or whitelist
```

File: src/ejabberd_sup.erl [hunk 12]
```diff
@@ -178,13 +178,20 @@ init([]) ->
            MucIQ,
            ShaperSup]}}.
 
+
```
Comment (chrzaszcz) [inline src/ejabberd_sup.erl:181]: Unnecessary extra space.

**Review by chrzaszcz — COMMENTED** [review]
Looks good, I added minor comments.

File: src/s2s/mongoose_s2s_lib.erl (hunk 11 above)
Comment (arcusfelis) [reply src/s2s/mongoose_s2s_lib.erl:148]: allowlist :)

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/181595) / elasticsearch_and_cassandra_mnesia / 936e2b165c43e539c8c9d93011df5a00fe5a1b4b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/181595/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/181595/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40b096eafdadd5.2023-07-20_12.47.49/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

File: src/config/mongoose_config_spec.erl [hunk 13]
```diff
@@ -172,6 +172,9 @@ general() ->
                  <<"sm_backend">> => #option{type = atom,
                                              validate = {module, ejabberd_sm},
                                              wrap = global_config},
+                 <<"s2s_backend">> => #option{type = atom,
+                                              validate = {module, mongoose_s2s},
+                                              wrap = global_config},
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/config/mongoose_config_spec.erl:179]: of course, if the used is weird. It will crash though when init function is called.
But it will pass the validation.
We could check behaviour type - but it is another story.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/181608) / small_tests / 1ee1ccc37d9636bbd74a1e6f90ec63a19ddc64db
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/181608/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/181608/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-07-20_13.49.48/lib.mongooseim.logs/run.2023-07-20_13.49.48/suite.log.html)

File: src/config/mongoose_config_spec.erl (hunk 13 above)
[the remark this answers is not in the document]
Comment (pawlooss1) [reply src/config/mongoose_config_spec.erl:179]: Ok let's leave it for now. It seems like too much effort for little improvement.

File: src/ejabberd_s2s_out.erl [hunk 14]
```diff
@@ -55,35 +55,62 @@
...
-                myname, server, queue,
+                from_to                 :: ejabberd_s2s:fromto(),
+                myname                  :: jid:lserver(),
+                server                  :: jid:lserver(),
+                queue                   :: element_queue(),
                 host_type               :: mongooseim:host_type(),
-                delay_to_retry = undefined_delay,
-                new = false             :: boolean(),
-                verify = false          :: false | {pid(), Key :: binary(), SID :: binary()},
+                delay_to_retry = undefined :: undefined | non_neg_integer(),
```
Comment (chrzaszcz) [inline src/ejabberd_s2s_out.erl:87]: Minor: `undefined` is the default value, so there is no need to explicitly assign it here.

File: src/ejabberd_s2s_out.erl [hunk 15]
```diff
@@ -329,44 +351,41 @@ wait_for_stream(closed, StateData) ->
...
-    case is_verify_res(El) of
-        {result, To, From, Id, Type} ->
+wait_for_validation({xmlstreamelement, El}, StateData = #state{from_to = FromTo}) ->
+    case mongoose_s2s_dialback:parse_validity(El) of
+        {step_3, FromTo, StreamID, IsValid} ->
+            ?LOG_DEBUG(#{what => s2s_receive_verify,
+                         from_to => FromTo, stream_id => StreamID, is_valid => IsValid}),
+            case StateData#state.verify of
+                false ->
+                    %% TODO: Should'nt we close the connection here ?
```
Comment (chrzaszcz) [inline src/ejabberd_s2s_out.erl:361]: Whenever I see a TODO like this in the code, I am wondering if you have simply forgotten to address it.
If not, I think it's better to state what you know than leave an open question, because IMO it is unlikely to be answered by anyone else.

Comment (chrzaszcz) [reply src/ejabberd_s2s_out.erl:361]:
```suggestion
                    %% TODO: Shouldn't we close the connection here ?
```

**Review by chrzaszcz — APPROVED** [review]
Looks good. For the future: please try to limit force pushes after review, especially for big PR's, because I had to re-read a lot of code for the third time... while what I'd like to see would be what changed after my last review, which is not easy to do.

**Review by chrzaszcz — APPROVED** [review]
Thanks it's all looking good now 👌

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/181835) / elasticsearch_and_cassandra_mnesia / 300b37c11d723e241e0df5b60d2e364af795cc3d
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4046/181835/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4046/181835/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40aa12dfcaaa77.2023-07-21_10.49.31/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0
